vi serve change detection fix

Raymi306 opened the pull request:
**IMPORTANT: Please do not create a Pull Request adding a new feature without discussing it first.**

The place to discuss new features is the forum: <https://zola.discourse.group/>
If you want to add a new feature, please open a thread there first in the feature requests section.

Sanity check:

* [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/getzola/zola/pulls) for the same update/change?

## Code changes
(Delete or ignore this section for documentation changes)

* [x] Are you doing the PR on the `next` branch?

If the change is a new feature or adding to/changing an existing one:

* [x] Have you created/updated the relevant documentation page(s)?

---

Fixes #2266 

A little bit of code duplication going on.
I made sure to leave a helpful comment explaining the extra branch.

Side note:  I noticed that notify library is 2 major versions behind.  Would you be interested in PRs for bumping major versions of dependencies, or don't fix what isn't broken?

Keats committed:
Next version

iEverX committed:
print error message when no config file found (#2168)

fixed #2195

MTRNord committed:
Speedup "zola check" command by reusing the Client (#2171)

* Reuse Client when checking urls and add timeout for requests

cydave committed:
Implement replace_re filter (#2163)

* Implement replace_re filter

* Cargo fmt

* add regex caching

* cargo fmt

* update docs, update unit test

* rename replace_re -> regex_replace

jk-ozlabs committed:
templates: add base URL for feed content (#2190)

Relative links in the entry content do not currently have a base URI, so
will be resolved relative to the feed URI:

Given an entry with the content:

    <a href="some-resource.bin">

And URIS of:

 * entry: https://example.org/blog/some-entry/
 * feed:  https://example.org/atom.xml

The link URI will end up as:

    https://example.org/some-resource.bin

rather than the URI that ends up resolved in the rendered page:

   https://example.org/blog/some-entry/some-resource.bin

The atom and RSS formats allow for an xml:base attribute (itself
specified in [1]) to provide a base URI of a subset of a document. This
change adds xml:base attributes to each entry, using the page permalink.

This gives us something equivalent to:

    <entry>
     <content xml:base="https://example.org/blog/some-entry/">
      <![CDATA[
       <a href="some-resource.bin">
      ]]>
     </content>
    </entry>

[1]: https://www.w3.org/TR/xmlbase/

Signed-off-by: Jeremy Kerr <jk@codeconstruct.com.au>

Jieiku committed:
Fix multi-lingual json index (#2197)

* Fix multi-ligual json index

* multi-lingual search Fix cargo fmt

Raymi306 committed:
Add search.index_format into the serialized config (#2165) (#2196)

* Add search into the serialized config (#2165)

* Only expose index_format

* Create config.search struct

* cargo fmt

Keats committed:
Fix typo in error message

Raymi306 committed:
Hard link serve panic fix (#2210)

* Fix hard link panic and add better error info to std:fs errors

* cargo fmt

* Remove erroneously committed config change

* Remove console import; Use with context to provide additional error info

* improve error wording

sinofp committed:
Add support for lazy loading images (#2211)

* Add optional decoding="async" loading="lazy" for img

In theory, they can make the page load faster and show content faster.

There’s one problem: CommonMark allows arbitrary inline elements in alt text.
If I want to get the correct alt text, I need to match every inline event.

I think most people will only use plain text, so I only match Event::Text.

* Add very basic test for img

This is the reason why we should use plain text when lazy_async_image is enabled.

* Explain lazy_async_image in documentation

* Add test with empty alt and special characters

I totaly forgot one can leave the alt text empty.
I thought I need to eliminate the alt attribute in that case,
but actually empty alt text is better than not having an alt attribute at all:
https://www.w3.org/TR/WCAG20-TECHS/H67.html
https://www.boia.org/blog/images-that-dont-need-alternative-text-still-need-alt-attributes
Thus I will leave the empty alt text.

Another test is added to ensure alt text is properly escaped.
I will remove the redundant escaping code after this commit.

* Remove manually escaping alt text

After removing the if-else inside the arm of Event::Text(text),
the alt text is still escaped.
Indeed they are redundant.

* Use insta for snapshot testing

`cargo insta review` looks cool!

I wanted to dedup the cases variable,
but my Rust skill is not good enough to declare a global vector.

TheOnlyMrCat committed:
Prevent spans crossing line boundaries in class-based code block formatter (#2237)

* Prevent spans crossing line boundaries in class formatter

* Add snapshot tests for classed highlighting

wold5 committed:
always sort assets by filename (Windows/Linux difference) (#2236)

* sort page.assets by filename

Uses .to_str() to sort files and subfolders.

The .unwrap() may need work or be replaced by unwrap_or_default(). Given
earlier checks in the function it should work however.

* add tests for assets sorting

* fix rustfmt

* use existing loop instead of windows

* also check the non-recursive test

* use .zip() and add assert msg

heitorPB committed:
Atom template authors (#2259)

* templates:atom: add support for multiple authors

Atom 1.0 [0] support multiple `<author>` entries in the feed. This commit
modified the template to generate as many `<author>` as the page's
metadata contains.

[0] https://validator.w3.org/feed/docs/atom.html#recommendedEntryElements

* Test we can have multiple authors in ATOM feeds

savq committed:
Add attributes to base_url link in atom template (#2261)

The `rel` and `type` HTML attributes are needed in the `base_url` (or
`section.permalink`) link so feed aggregators know that's the HTML page
that corresponds to the atom feed.

Note: The RSS template doesn't have this issue.

Raymi306 committed:
Fixes #2250; Error instead of panic when root directory or config file not present (#2262)

adeadfed committed:
Fix LFI in `zola serve` (#2258)

* use fs canonicalize to prevent path traversal

* fix cargo fmt

Keats committed:
Update changelog

Keats committed:
Update edition

Keats committed:
Update deps

Raymi306 committed:
Add ignored_static to config (#2209)

* Add ignored_static to config

* Make  handle ignored static files correctly

* cargo fmt

* Match on relative path rather than incorrect target path

* path -> partial path for serve static ignore

* remove debug println

* copy static directory if there is no ignored globset

* Update docs

* Deduplicate code with additional Option argument

* cargo fmt

heitorPB committed:
template:feeds: add extra block (#2263)

* template:feeds: add extra block

* add missing >

* Revert "add missing >"

This reverts commit 45c6b9c6eb0d9964fa6c2b4fcc7152c1b31bf1d6.

* Revert "template:feeds: add extra block"

This reverts commit 596f7f1e7f3b221ad06cbde557e06ec4980e166f.

* Update docs for feed templates

Raymi306 committed:
Add fix for detecting changes on files by editors which replace files on save, such as vi

Keats commented on src/cmd/serve.rs:
```diff
@@ -653,6 +653,29 @@ pub fn serve(
                         messages::report_elapsed_time(start);
                     }
+                    NoticeRemove(path) => {
+                        // Some editors, like those in the vi family, replace the file you're
```
vim works for me, is it only vi having issues?

iamorphen replied:
On macOS, hot reloading works with vim/nvim for me. On Linux, neither work for me with `zola` built from master.

Raymi306 committed:
Attempt to prevent possible race condition

If the site rebuilds and the config.toml file hasn't been moved over on top of the removed version, the site cannot rebuild correctly
